File: Res/Sorting/BucketSort/bucketSort.dart
void bucketSort(List<int> list) {
  int n = list.length;
  if (n <= 0) return;

  // 1. Find the maximum value in the list to normalize the elements
  int maxValue = list.reduce((a, b) => a > b ? a : b);

  // 2. Create a empty buckets
  List<List<int>> buckets = List.generate(n, (index) => []);

  // 3. Distributed elements into buckets
  for (int i = 0; i < n; i++) {
    int bucket_index = (list[i] * n ~/ (maxValue + 1));
    buckets[bucket_index].add(list[i]);
  }
  // 4. Sort individual buckets
  for (int i = 0; i < n; i++) {
    buckets[i].sort();
  }
  // 5. Concatenate all sorted buckets
  int index = 0;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < buckets[i].length; j++) {
      list[index++] = buckets[i][j];
    }
  }
}

void main() {
  List<int> list = [25, 10, 47, 95, 62, 30, 75, 45];
  bucketSort(list);
  print(list);
}

File: Res/Sorting/BucketSort/bucketSort.py
def bucketSort(arr):
    n = len(arr)
    if n <= 1:
        return
    
    max_value = max(arr)
    
    bucket = [[] for _ in range(n)]
    
    for i in range(n):
        bucket_index = arr[i] * n // (max_value + 1)
        bucket[bucket_index].append(arr[i])
        
    for i in range(n):
        bucket[i].sort()
    
    index = 0
    for i in range(n):
        for j in range(len(bucket[i])):
            arr[index] = bucket[i][j]
            index += 1

arrs = [25, 10, 47, 95, 62, 30, 75, 45]
bucketSort(arrs)
print(arrs)